samples: net: add mutual DTLS to download and coap_client (#30295)
nrfconnect/sdk-nrf: samples/net/coap_client/README.rst

The CoAP Client sample performs the following actions:

#. Connect to the configured public CoAP test server (specified by the Kconfig option :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME`).

> #. Send periodic GET request for a test resource (specified by the Kconfig option :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE`) that is available on the server.

Reviewer: ```suggestion
#. Connect to the configured public CoAP test server (specified by the :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME` Kconfig option).
#. Send a periodic GET request for a test resource (specified by the :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE` Kconfig option) that is available on the server.
```

> #. Send a periodic GET request for a test resource (specified by the :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE` Kconfig option) that is available on the server.

Reviewer 2: ```suggestion
#. Connect to the configured public CoAP test server (specified by the :option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME` Kconfig option).
#. Send a periodic GET request for a test resource (specified by the :option:`CONFIG_COAP_SAMPLE_RESOURCE` Kconfig option) that is available on the server.
```

> #. Send a periodic GET request for a test resource (specified by the :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE` Kconfig option) that is available on the server.

Reviewer 3: Maybe remove the kconfig mentioning from here? They are anyway mentioned in the subsection below (`Selecting the server and resource`).

This sample uses the resource **obs** (Californium observable resource) in the communication between the CoAP client and the public CoAP server.
The communication follows the standard request/response pattern and is based on the change in the state of the value of the resource.
The sample queries one resource at a time.

Selecting the server and resource
==================================

The server and resource to query are configured with Kconfig options and default to the Californium test server:

* :option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME` and :option:`CONFIG_COAP_SAMPLE_SERVER_PORT`- The CoAP server to connect to.
* :option:`CONFIG_COAP_SAMPLE_RESOURCE`- The resource to query on that server.

.. _coap_client_sample_mtls:

Mutual DTLS (client certificate authentication)
===============================================

The sample can optionally use DTLS to secure the CoAP communication (CoAPS), with client authentication (mutual TLS).

> Note that this is only supported on Wi-Fi boards and not on cellular boards.

Reviewer 2: ```suggestion
.. note::
   This functionality is only supported on Wi-Fi boards and not on cellular boards.
```

This sample uses the resource **obs** (Californium observable resource) in the communication between the CoAP client and the public CoAP server.
The communication follows the standard request/response pattern and is based on the change in the state of the value of the resource.
The sample queries one resource at a time.

Selecting the server and resource
==================================

> The server and resource to query are configured with Kconfig options, and default to the Californium test server described above:

Reviewer 2: ```suggestion
The server and resource to query are configured with Kconfig options and default to the Californium test server:
```

> * :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME` and :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_PORT`—the CoAP server to connect to.

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME` and :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_PORT`- the CoAP server to connect to.
```

> * :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE`—the resource to query on that server.

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE`- the resource to query on that server.
```

> * :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE`- the resource to query on that server.

Reviewer 2: ```suggestion
* :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_HOSTNAME` and :kconfig:option:`CONFIG_COAP_SAMPLE_SERVER_PORT`- The CoAP server to connect to.
* :kconfig:option:`CONFIG_COAP_SAMPLE_RESOURCE`- The resource to query on that server.
```

The sample can optionally use DTLS to secure the CoAP communication (CoAPS), with client authentication (mutual TLS).
Enable the :kconfig:option:`CONFIG_COAP_SAMPLE_DTLS` option to provision a CA certificate, a client certificate, and a client private key, and to connect over CoAPS using mutual X.509 authentication.
Set :kconfig:option:`CONFIG_COAP_SAMPLE_CA_CERT_FILE`, :kconfig:option:`CONFIG_COAP_SAMPLE_CLIENT_CERT_FILE`, and :kconfig:option:`CONFIG_COAP_SAMPLE_CLIENT_KEY_FILE` to the CA certificate, client certificate, and client private key to provision, matching what the server you connect to expects.

The sample includes an example CA trust chain and client certificate and private key under :file:`cert/`, for use against the `Eclipse Californium`_ CoAP interop server:

* :file:`cert/cf-ca.pem` — CA trust chain, used to validate the server certificate
* :file:`cert/cf-client.pem` — client certificate
* :file:`cert/cf-client-key.pem` — client private key (EC P-256)

The :file:`wifi-dtls.conf` extra-conf file configures the sample to use these to connect to ``coaps://californium.eclipseprojects.io:5684`` using mutual X.509 authentication (ECDHE-ECDSA-AES128-GCM-SHA256).

> .. _Eclipse Californium: https://github.com/eclipse-californium/californium

Reviewer 2: Please add this link to the links.txt - https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/links.txt

Wi-Fi
=====

> On Wi-Fi boards, networking and TLS/DTLS support are not part of the default configuration and must be added with the :file:`wifi.conf` extra-conf file, using the ``coap_client_EXTRA_CONF_FILE`` sysbuild variable.

Reviewer 3: Having `and TLS/DTLS support` here is confusing as `wifi.conf` doesn't enable DTLS, right? Same for `download` sample.

Author: Good catch. Removing now for `coap_client`.

For `download` it is different: that sample had built-in TLS in its default config and now DTLS is also built in. In that sample, `dtls-test.conf` really just points to DTLS resources, assuming DTLS is already enabled.

For example, to build for the nRF7120 DK using plain CoAP::

   west build -p -b nrf7120dk/nrf7120/cpuapp/ns --sysbuild samples/net/coap_client -- \
     -Dcoap_client_EXTRA_CONF_FILE="wifi.conf"

> To instead perform mutual DTLS (CoAPS) against the Californium interop server (see :ref:`Mutual DTLS (client certificate authentication) <coap_client_sample_mtls>` above), add the :file:`wifi-dtls.conf` extra-conf file on top of :file:`wifi.conf`::

Reviewer 2: ```suggestion
To perform mutual DTLS (CoAPS) with the Californium interoperability server (see :ref:`Mutual DTLS (client certificate authentication) <coap_client_sample_mtls>`), add the :file:`wifi-dtls.conf` extra-conf file on top of :file:`wifi.conf`::
```

   west build -p -b nrf7120dk/nrf7120/cpuapp/ns --sysbuild samples/net/coap_client -- \

>      -Dcoap_client_EXTRA_CONF_FILE="wifi.conf;wifi-dtls.conf"

Reviewer 3: Perhaps we should follow this guide: https://github.com/nrfconnect/sdk-nrf/blob/63ffa5966b96c3e6aec9c003f1321b197caeeed5/doc/nrf/templates/sample_README.rst?plain=1#L157-L170 and just mention these .conf files without building command? My concern is that we have to maintain these `west build ...` commands while there is a common place that explains how to use them: https://ncsbmdoc.z6.web.core.windows.net/ncs/PR-30295/nrf/app_dev/config_and_build/cmake/index.html

Same for `download` sample.

Author: Agree
